File: 922.Sort-Array-By-Parity-II.go
// https://leetcode.com/problems/evaluate-division/description/
//
// algorithms
// Easy (68.13%)
// Total Accepted:    6.2k
// Total Submissions: 9.1k
// beats 56.52% of golang submissions

package leetcode

func sortArrayByParityII(A []int) []int {
	length := len(A)

	for i := 0; i < length; i++ {
		if i%2 == 0 && A[i]%2 == 1 {
			for j := i + 1; j < length; j++ {
				if A[j]%2 == 0 {
					A[i], A[j] = A[j], A[i]
					break
				}
			}
		} else if i%2 == 1 && A[i]%2 == 0 {
			for j := i + 1; j < length; j++ {
				if A[j]%2 == 1 {
					A[i], A[j] = A[j], A[i]
					break
				}
			}
		}
	}

	return A
}

File: 922.Sort-Array-By-Parity-II.java
// https://leetcode.com/problems/sort-array-by-parity-ii/
//
// algorithms
// Easy (67.58%)
// Total Accepted: 68,801
// Total Submissions: 101,803
// beats 100.0% of java submissions


class Solution {
    public int[] sortArrayByParityII(int[] A) {
        boolean isOdd = false;
        int pre = 0;
        int len = A.length;
        int idx = 0;

        while (idx < len) {
            if (isOdd) {
                if ((A[idx] & 1) == 1) {
                    swap(A, idx, pre);
                    pre++;
                    isOdd = !isOdd;
                } else {
                    idx++;
                }
            } else {
                if ((A[idx] & 1) == 0) {
                    swap(A, idx, pre);
                    pre++;
                    isOdd = !isOdd;
                } else {
                    idx++;
                }
            }
        }

        return A;
    }

    private void swap(int[] A, int i, int j) {
        int tmp = A[i];
        A[i] = A[j];
        A[j] = tmp;
    }
}